synapse/handlers/sync.py

```diff
@@ -571,15 +571,18 @@ def _generate_sync_entry_for_to_device(self, sync_result_builder):
                 user_id, device_id, since_stream_id
             )
 
-        logger.debug("Getting messages up to %d", now_token.to_device_key)
-        messages, stream_id = yield self.store.get_new_messages_for_device(
-            user_id, device_id, now_token.to_device_key
-        )
-        logger.debug("Got messages up to %d: %r", stream_id, messages)
-        sync_result_builder.now_token = now_token.copy_and_replace(
-            "to_device_key", stream_id
-        )
-        sync_result_builder.to_device = messages
+        if since_stream_id and since_stream_id == int(now_token.to_device_key):
```

Comment:
I'm not sure about the ``since_stream_id`` check here either. I guess you are trying to avoid setting the stream_id to nonzero and therefore triggering the ``delete_messages_for_device`` code above?